hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/BootstrapNodeManager.java

```diff
@@ -0,0 +1,218 @@
+  public static final long DEFAULT_REQUEST_MASTER_INTERVAL_SECS = TimeUnit.MINUTES.toSeconds(10);
+
+  public static final String REQUEST_MASTER_MIN_INTERVAL_SECS =
+    "hbase.server.bootstrap.request_master_min_interval.secs";
+
+  // default 30 seconds
+  public static final long DEFAULT_REQUEST_MASTER_MIN_INTERVAL_SECS = 30;
+
+  public static final String REQUEST_REGIONSERVER_INTERVAL_SECS =
+    "hbase.server.bootstrap.request_regionserver_interval.secs";
+
+  // default request every 30 seconds
+  public static final long DEFAULT_REQUEST_REGIONSERVER_INTERVAL_SECS = 30;
+
+  private static final float JITTER = 0.2f;
+
+  private volatile List<ServerName> nodes = Collections.emptyList();
+
+  private final AsyncClusterConnection conn;
+
+  private final MasterAddressTracker masterAddrTracker;
+
+  private final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor(
+    new ThreadFactoryBuilder().setDaemon(true).setNameFormat(getClass().getSimpleName()).build());
+
+  private final long requestMasterIntervalSecs;
+
+  private final long requestMasterMinIntervalSecs;
+
+  private final long requestRegionServerIntervalSecs;
+
+  private final int maxNodeCount;
+
+  private final RetryCounterFactory retryCounterFactory;
+
+  private RetryCounter retryCounter;
+
+  private long lastRequestMasterTime;
+
+  public BootstrapNodeManager(AsyncClusterConnection conn, MasterAddressTracker masterAddrTracker) {
+    this.conn = conn;
+    this.masterAddrTracker = masterAddrTracker;
+    Configuration conf = conn.getConfiguration();
+    requestMasterIntervalSecs =
+      conf.getLong(REQUEST_MASTER_INTERVAL_SECS, DEFAULT_REQUEST_MASTER_INTERVAL_SECS);
+    requestMasterMinIntervalSecs =
+      conf.getLong(REQUEST_MASTER_MIN_INTERVAL_SECS, DEFAULT_REQUEST_MASTER_MIN_INTERVAL_SECS);
+    requestRegionServerIntervalSecs =
+      conf.getLong(REQUEST_REGIONSERVER_INTERVAL_SECS, DEFAULT_REQUEST_REGIONSERVER_INTERVAL_SECS);
+    maxNodeCount = conf.getInt(HBaseRpcServicesBase.CLIENT_BOOTSTRAP_NODE_LIMIT,
+      HBaseRpcServicesBase.DEFAULT_CLIENT_BOOTSTRAP_NODE_LIMIT);
+    retryCounterFactory = new RetryCounterFactory(
+      new RetryConfig().setBackoffPolicy(new ExponentialBackoffPolicyWithLimit()).setJitter(JITTER)
+        .setSleepInterval(requestMasterMinIntervalSecs).setMaxSleepTime(requestMasterIntervalSecs)
+        .setTimeUnit(TimeUnit.SECONDS));
+    executor.schedule(this::getFromMaster, getDelay(requestMasterMinIntervalSecs),
+      TimeUnit.SECONDS);
+  }
+
+  private long getDelay(long delay) {
+    long jitterDelay = (long) (delay * ThreadLocalRandom.current().nextFloat() * JITTER);
+    return delay + jitterDelay;
+  }
+
+  private void getFromMaster() {
+    List<ServerName> liveRegionServers;
+    try {
+      // get 2 times number of node
+      liveRegionServers =
+        FutureUtils.get(conn.getLiveRegionServers(masterAddrTracker, maxNodeCount * 2));
+    } catch (IOException e) {
```

Comment:
If master is down for a while, we can't get bootstrap nodes from other regionservers, But it seems not a problem.

Reply:
Not a big problem, if client cold contact us for getting bootstrap node, at least it could also call other methods on us :)